python/ray/data/_internal/planner/plan_udf_map_op.py

```diff
@@ -193,45 +220,97 @@ def _validate_batch_output(batch: Block) -> None:
-                    # calling the UDF.
-                    # TODO(hchen): This workaround is because some all-to-all
-                    # operators output empty blocks with no schema.
-                    res = [batch]
+    if inspect.iscoroutinefunction(fn):
+        # UDF is a callable class with async generator `__call__` method.
+        def transform_fn(
+            input_iterable: Iterable[DataBatch], _: TaskContext
+        ) -> Iterable[DataBatch]:
+            # Use a queue to store outputs from async generator calls.
+            # We will put output batches into this queue from async
+            # generators, and in the main event loop, yield them from
+            # the queue as they become available.
+            output_batch_queue = queue.Queue()
+
+            async def process_batch(batch: DataBatch):
+                output_batch_iterator = await fn(batch)
+                # As soon as results become available from the async generator,
+                # put them into the result queue so they can be yielded.
+                async for output_batch in output_batch_iterator:
+                    output_batch_queue.put(output_batch)
+
+            async def process_all_batches():
+                loop = ray.data._cached_loop
+                tasks = [loop.create_task(process_batch(x)) for x in input_iterable]
+
+                ctx = ray.data.DataContext.get_current()
+                if ctx.execution_options.preserve_order:
+                    for task in tasks:
+                        await task()
                 else:
-                    res = fn(batch)
-                    if not isinstance(res, GeneratorType):
-                        res = [res]
-            except ValueError as e:
-                read_only_msgs = [
-                    "assignment destination is read-only",
-                    "buffer source array is read-only",
-                ]
-                err_msg = str(e)
-                if any(msg in err_msg for msg in read_only_msgs):
-                    raise ValueError(
-                        f"Batch mapper function {fn.__name__} tried to mutate a "
-                        "zero-copy read-only batch. To be able to mutate the "
-                        "batch, pass zero_copy_batch=False to map_batches(); "
-                        "this will create a writable copy of the batch before "
-                        "giving it to fn. To elide this copy, modify your mapper "
-                        "function so it doesn't try to mutate its input."
-                    ) from e
+                    for task in asyncio.as_completed(tasks):
+                        await task
+                # Sentinel to indicate completion.
+                output_batch_queue.put(None)
+
+            # Use the existing event loop to create and run
+            # Tasks to process each batch
+            loop = ray.data._cached_loop
+            loop.run_until_complete(process_all_batches())
```

Comment:
This is still running in the main thread. Thus cannot run multiple batches at the same time.
We should put `loop.run_forever()` in a thread, and then call `loop.call_soon_threadsafe` here.